Simplenote/SPHistoryLoader.swift

```diff
@@ -0,0 +1,109 @@
+import Foundation
+
+// MARK: - SPHistoryLoader: Request and aggregate versions of Simperium object
+//
+final class SPHistoryLoader {
+
+    // MARK: - Item: Represents a version object
+    //
+    struct Item {
+        let version: Int
+        let data: [String: Any]
+    }
```

Comment:
Suggestion: Both `SPHistoryLoader.Item` and `SPNoteHistoryController.Presentable` end up encapsulating the same entity: a Note Version.

How about having a single non nested entity, and parsing the fields you need?
That way you can avoid parsing the data collection in multiple spots

Reply:
I agree about having an object with strongly typed fields instead of just having a dictionary. I can definitely use `Codable` here.

But I think it is beneficial to have different entities. Even though they're both encapsulating a note version, the purpose and the fields are different: `SPHistoryLoader.Item` represents data that is received from Simperium and `SPNoteHistoryController.Presentable` is created from `SPHistoryLoader.Item` with only fields that are required to render the UI. For example, they both have `date` field, but `SPNoteHistoryController.Presentable` store it already formatted according to UI needs.

In other words, If I would create a framework for Simplenote to separate data management from UI, `SPHistoryLoader` would go to the framework, but `SPNoteHistoryController` would stay in the app as it is part of UI and transforms framework data to the format suitable for the user.

Reply:
I've removed the dictionary from `SPHistoryLoader.Item` in https://github.com/Automattic/simplenote-ios/pull/795/commits/708dab0155e1c014b087ccf78f62b6568248e4f5 and I think it became better now 👍

Reply:
I do appreciate separation of concerns, **but** (for the time being, at least) splitting into a Framework is not on the pipeline. 

Keeping a single entity would be beneficial because:

1. It'd be easier to read (`Item` / `Presentable` are super generic, though namespaced)
2. The codebase would shrink (no need to map Item > Presentable, less code overall)
3. The `Int(note.version() ?? "1") ?? 1` snippet could go in a single spot
4. And both, Item and Presentable represent the same entity

What do you think?

Reply:
I didn't mean that there is a need for a framework, but I think that it's nice to separate things.

> It'd be easier to read (Item / Presentable are super generic, though namespaced)

I can move `Item` out of `SPHistoryLoader`, but I think `Presentable` is best being namespaced as it's only purpose it to convey information from history controller to history view controller and won't be used by any other code.

> The codebase would shrink (no need to map Item > Presentable, less code overall)

I don't believe it will as there is still need to map history Item to UI. In case of using one entity, the code to format date and check if note's version is current will move to the view controller.

> The Int(note.version() ?? "1") ?? 1 snippet could go in a single spot

This actually would be nice to have in an extension of Note. But either way we would still need to access it from both loader (to check available versions) and in a view controller (to check if selected version is current)

> And both, Item and Presentable represent the same entity

I would say it's a different representation of the same entity. One is how it is handled by Simperium API and the other one is a representation for the UI.

I think the difference is not that big and by not using `Presentable` we shift code from controller to view controller, but is it a good thing?